test_tf2/test/permuter.hpp

```diff
@@ -0,0 +1,169 @@
+ *       this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+/** \author Tully Foote */
+
+#ifndef ROSTEST_PERMUTER_H
+#define ROSTEST_PERMUTER_H
+
+#include <vector>
+#include <mutex>
+
+namespace rostest
+{
+/** \brief A base class for storing pointers to generic data types
+ */
+class PermuteOptionBase
+{
+public:
+  virtual void reset() =0;
+  virtual bool step() =0;
+  virtual ~PermuteOptionBase() {};
+};
+
+
+/**\brief A class to hold a set of option values and currently used state
+ *  This class holds
+ */
+template<class T>
+class PermuteOption : public PermuteOptionBase
+{
+public:
+  PermuteOption(const std::vector<T>& options, T* output)
+  {
+    options_ = options;
+    output_ = output;
+    reset();
+  }
+
+  virtual ~PermuteOption(){};
+
+  void reset(){
+    std::lock_guard<std::mutex> lock(access_mutex_);
+    current_element_ = options_.begin();
+    *output_ = *current_element_;
+  };
+
+  bool step()
+  {
+    std::lock_guard<std::mutex> lock(access_mutex_);
+    current_element_++;
+    if (current_element_ == options_.end())
+      return false;
+    *output_ = *current_element_;
+    return true;
+  };
+
+private:
+  /// Local storage of the possible values
+  std::vector<T> options_;
+  /// The output variable
+  T* output_;
+  typedef typename std::vector<T>::iterator V_T_iterator;
+  /// The last updated element
+  V_T_iterator current_element_;
+
+  std::mutex access_mutex_;
+
+};
+
+/** \brief A class to provide easy permutation of options
+ * This class provides a way to collapse independent
+ * permutations of options into a single loop.
+ */
+class Permuter
+{
+public:
+  /** \brief Destructor to clean up allocated data */
+  virtual ~Permuter(){ clearAll();};
+
+
+  /** \brief Add a set of values and an output to the iteration
+   * @param values The set of possible values for this output
+   * @param output The value to set at each iteration
+   */
+  template<class T>
+  void addOptionSet(const std::vector<T>& values, T* output)
+  {
+    std::lock_guard<std::mutex> lock(access_mutex_);
+    options_.push_back(static_cast<PermuteOptionBase*> (new PermuteOption<T>(values, output)));
```

Comment:
Instead of using `push_back` and `new` here, you can use `emplace_back`.  Also, it would be slightly nicer to use a vector of `std::unique_ptr` (that way you don't have to do the cleanup in the destructor).